### PR TITLE
topdown: Add new numbers.range built-in function

### LINKED ISSUE #2479: Add built-in function to generate range of numbers between min and max
A few recent use cases could have used a built-in function to generate a range of integers between two numbers. For example:

```rego
a = -1
b = 3
numbers.range(a, b) == [-1,0,1,2,3]
numbers.range(a, a) == [-1]
numbers.range(b, a) == [3,2,1,0,-1]
```

Note, since OPA supports arbitrary-precision numbers, we should use big.Int for the implementation.

### PR DESCRIPTION
**Pull request by tsandall**
This commit adds a new built-in function to generate a range of
integers between two values (inclusive). This is useful in certain
cases where users need to enumerate a set of values (e.g., port
numbers).

Fixes #2479

Signed-off-by: Torin Sandall <torinsandall@gmail.com>

### CODE REVIEW

File: docs/content/policy-reference.md [hunk 1]
```diff
@@ -357,6 +357,12 @@ complex types.
 | <span class="opa-keep-it-together">``output := trim_space(string)``</span> | ``output`` is a ``string`` representing ``string`` with all leading and trailing white space removed.|
 | <span class="opa-keep-it-together">``output := upper(string)``</span> | ``output`` is ``string`` after converting to upper case |
 
+### Numbers
+
+| Built-in | Description |
+| ------- |-------------|
+| <span class="opa-keep-it-together">``output := numbers.range(a, b)``</span> | ``output`` is the range of integer numbers between ``a`` and ``b`` (inclusive). If ``a`` == ``b`` then ``output`` == ``[a]``. If ``a`` < ``b`` the range is in ascending order. If ``a`` > ``b`` the range is in descending order. |
+
```
Comment (ashutosh-narkar) [inline docs/content/policy-reference.md:365]: Nit: `output = [a]`.

We already have a `Numbers` section in the Built-in function docs. Let's either combine or rename this.

File: topdown/numbers_test.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+		{
+			note: "descending",
+			stmt: "p = x { x := numbers.range(2, -3) }",
+			exp:  "[2, 1, 0, -1, -2, -3]",
+		},
+		{
+			note: "precision",
+			stmt: "p { numbers.range(49649733057, 49649733060, [49649733057, 49649733058, 49649733059, 49649733060]) }",
+			exp:  "true",
+		},
```
Comment (patrick-east) [inline topdown/numbers_test.go:35]: We've been adding in negative tests, in particular for the arguments, should we add a few here too? If anything just catching like if a float number is passed in.

File: docs/content/policy-reference.md (hunk 1 above)
Comment (tsandall) [reply docs/content/policy-reference.md]: Merged this into the existing one. Forgot we had it.

File: topdown/numbers.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
+// Copyright  The OPA Authors.  All rights reserved.
```
[the remark this answers is not in the document]
Comment (tsandall) [reply topdown/numbers.go]: Fixed.

File: topdown/numbers_test.go (hunk 2 above)
Comment (tsandall) [reply topdown/numbers_test.go:35]: Added tests for floating point error condition.

**Comment by tsandall** [comment]
@patrick-east I've updated this PR to include the wasm version as well. PTAL.

**Comment by tsandall** [comment]
@patrick-east updated the PR. I think I mentioned offline a little while ago that the built-in function name => wasm library function name mapping should live inside the wasm backend as opposed to the planner...I've gone ahead and refactored so that's the case. PTAL.

**Pull request merged**